keps/sig-node/20181106-in-place-update-of-pod-resources.md

```diff
@@ -172,9 +184,24 @@ available memory are more probable to require restarts.
 If more than one resource type with different policies are updated, then
 RestartContainer policy takes precedence over NoRestart policy.
 
-Additionally, if RestartPolicy is 'Never', ResizePolicy should be set to
+Additionally, if RestartPolicy is 'Never', ResizePolicy must be set to
```

Comment:
Actually - is it even possible to do an in-place resize of a "Never" pod?  If we can't guarantee that we can do the resize without a restart, the result of such a resize would be very non-deterministic.  Sometimes it works and sometimes it kills the pod completely.

Reply:
Yes, but only with ResizePolicy=NoRestart. A RestartPolicy of Never means don't restart container regardless of why it exited. It is at odds with ResizePolicy=Restart. So for NoRestart + Never, kubelet at this time will keep retrying UpdateContainerResources CRI until it succeeds, it won't stop and start the container to get it resize.

Reply:
Now you've muddied the semantics again.

If I set all on my resources' policies to "NoRestart", is it GUARANTEED to not restart?  Or is it valid to restart anyway for some internal reason?

This is a really important semantic to pin down.  It's easy to imagine a RuntimeClass that can support resize but only with a  restart, or even sometimes with a restart.  A VM-ish runtime might be able to grow within an integral core, but not bring new cores online.  E.g. I have 1.5 cores and I grow to 1.8 cores => no restart needed, but 1.5 -> 2.1 would need a restart to bring that 3rd CPU instance on-line in the guest.

So please help me understand the exact meaning of this. :)

Reply:
Kubelet guarantees not to restart the container, there is no internal reason to restart (StopContainer + StartContainer CRIs).

In above case, user needs to set policy to Restart at this time. And policy mechanism would need to add support for sub-integral grow/shrink.

Reply:
I combed through kubelet SyncPod function again to ensure I didn't miss a reason for us to restart despite NoRestart policy - there isn't any. If the container stops on its own volition in response to our UpdateContainerResources call, that's not on us.

We can update the documentation in the code to say we will attempt resizing the container it without restarting it. (Remove the 'if possible' wording). If we need an option like 'NoRestartIfPossible' at later point, we can always add it.

Reply:
I'm not convinced this is the sanest semantic.

For the sake of argument, can we assume a VM-ish runtime as I described above?  It can sometimes resize CPU without restarting but not always.  The exact details of that are not important, but does it seem at least plausible?

I have a Pod YAML that says the CPU resizePolicy is "NoRestart".  I know that the application in question does not *need* to be Restarted.  I probably got this Pod from some helm chart that someone else authored - it's not my own creation.  

On a "normal" runtime (e.g. containerd), this pod is fine.  But what about this hypothetical runtime?  Let's assume the pod can be created.

Now I trigger a resize action.

Again, on a "normal" runtime, this will be fine.  But on this hypothetical runtime,what happens?

a) The resize operation fails statically - this runtime doesn't support resizes with NoRestart (sometimes it needs to restart)

b) The resize operation is accepted but never completes - the runtime can't actuate it without a restart

c) The resize operation is accepted and actuated, causing a restart if needed

In cases (a) and (b), we broke portability.  I have to change my pod YAML based on my runtime.  IMO, that's about as bad as it gets.  I think (c) is the only viable course.  That means the semantics of "NoRestart" are "This container does not need to be restarted, but the runtime may restart it anyway".

"restart (not required)" != "(not restart) required"

So to my own original comment, a pod that says "restart=Never" simply can not safely be resized, because a resize might cause a restart (unless we add a policy that expresses the alternative mode.

Reply:
From the user perspective, you're correct. What runtime does in response to UpdateContainerResources CRI call is a concern, and beyond our scope of influence. From kubelet's perspective, NoRestart means kubelet will not stop & start the container - we have control over that, and that's what we can guarantee.

resizePolicy of NoRestart certainly gives the wrong impression considering this possibility. ```LiveUpdate``` (or do you prefer ```LiveResize```?) seems to better express what this policy value implies. (Thanks for picking on this point!)

However, do you see a reason why we cannot enforce in validation that restartPolicy=Never requires resizePolicy==LiveUpdate? Because Restart or potential future RestartOnGrow/Shrink means we know we will be restarting the container at some point. To me it sounds like a useful sanity check.

So, to nail this down better, besides going with the term ``LiveUpdate```, how does the following sound:

---
To provide fine-grained user control, PodSpec.Containers is extended with ResizePolicy - a list of named subobjects (new object) that supports 'cpu' and 'memory' as names. It supports the following policy values:
  * LiveUpdate - default value; resize the Container while continues to run,
  * Restart - restart the Container in-place to apply new resource values. (e.g. Java process needs to change its Xmx flag)

By using ResizePolicy, user can mark Containers as safe (or unsafe) for in-place resource update. Kubelet uses it to determine the required action.

LiveUpdate policy does not guarantee that container won't restart. The runtime may choose to stop the container if it is unable to apply the new resources without needing restart. If RestartPolicy of 'Never' is used, in-place resize may result in the container not being started, should it stop during resize.

Reply:
> NoRestart means kubelet will not stop & start the container - we have control over that, and that's what we can guarantee

I think that's a cop-out and likely to make for very angry users when some runtime DOES do a restart.  If we think it's VALID for a runtime to demand a restart, we should just say so in the docs.  If we think it is INVALID, then we should make that very clear, and exclude such runtimes from participating in Kubernetes.

Your edited text looks good to me, execpt:



```
- If RestartPolicy of 'Never' is used, in-place resize may result in the container not being started, should it stop during resize.
+ If RestartPolicy of 'Never' is used, in-place resize may result in the container being stopped and not restarted, if the system can not perform the resize without a restart.
```

Reply:
I'll discuss this in sig-node meeting tomorrow - Mike Brown who maintains containerd may have some valuable insights.

Excluding runtimes who cannot honor restart-free resize is ideal goal but I'm not sure if there's a bullet-proof way we can test for it to guarantee that k8s-certified runtime will never someday stop the container if it can't handle UpdateContainerResourcesRequest. (I'm presuming that's the case we are worried about here)